modyn/supervisor/internal/triggers/performance/performance_tracker.py

```diff
@@ -0,0 +1,134 @@
+    evaluations since the last trigger. When a trigger happens, this
+    series of observations evaluations is reset.
+
+    Provides both the wrapped performance metrics as well as accuracy
+    information.
+    """
+
+    def __init__(self, trigger_eval_window_size: int) -> None:
+        """
+        Args:
+            window_size: How many evaluations after triggers should be kept in memory.
+        """
+        self.trigger_evaluation_memory: deque[tuple[int, int, dict[str, float]]] = deque(
+            maxlen=trigger_eval_window_size
+        )
+        """Memory of the last `window_size` evaluations after triggers with
+        their number of samples, misclassifications and evaluation scores for
+        different metrics.
+
+        After every trigger, the memory is updated with the new
+        evaluation.
+        """
+
+        self.since_last_trigger: list[tuple[int, int, dict[str, float]]] = []
+        """Memory of the evaluations since the last trigger with their number
+        of samples, misclassifications and evaluation scores for different
+        metrics.
+
+        Upon trigger, this memory is reset.
+        """
+
+    def inform_evaluation(
+        self,
+        num_samples: int,
+        num_misclassifications: int,
+        evaluation_scores: dict[str, float],
+    ) -> None:
+        """Informs the tracker about a new evaluation."""
+        self.since_last_trigger.append((num_samples, num_misclassifications, evaluation_scores))
+
+    def inform_trigger(
+        self,
+        num_samples: int,
+        num_misclassifications: int,
+        evaluation_scores: dict[str, float],
+    ) -> None:
+        """Informs the tracker about a new trigger and restarts the tracker of
+        expected performance evaluations.
+
+        The expected performance evaluations are the evaluations right after triggers (`since_last_trigger`).
+        """
+        self.trigger_evaluation_memory.append((num_samples, num_misclassifications, evaluation_scores))
+
+        # reset the since_last_trigger tracker: first element in the new series is the performance right after trigger
+        self.since_last_trigger = [(num_samples, num_misclassifications, evaluation_scores)]
+
+    @property
+    def previous_batch_num_misclassifications(self) -> int:
+        """Returns the number of misclassifications in the previous batch."""
+        return self.since_last_trigger[-1][1]
+
+    def forecast_expected_accuracy(self, method: ForecastingMethod = "ridge_regression") -> float:
+        """Forecasts the expected accuracy based on the current memory of
+        evaluations right after triggers.
+
+        Returns:
+            The forecasted accuracy.
+        """
+        return forecast_next_time_series_value(
+            observations=[1 - p[1] / p[0] for p in self.trigger_evaluation_memory],
+            method=method,
+        )
+
+    def forecast_next_accuracy(self, method: ForecastingMethod = "ridge_regression") -> float:
```

Comment:
I think my question about this is why we need it. We already enforce the user to use Accuracy, hence we could use  `forecast_expected_performance` with accuracy instead of "manually" calculating the accuracy here?

Reply:
I see your point; generally, I'd prefer your solution. However, this assumption that we force the user to specify accuracy is on the configuration level. For this component, I'd rather keep it independent of this assumption and use stick with the duality of `performance` / `accuracy.`